### PR TITLE
Bugzilla plugin is now converted to backend sync service

### PR DESCRIPTION
**Pull request by tkdchen**
as the summary, Bugzilla is now a backend sync service. It can be used simply as it was a plugin before. When enable this service, any added task will be handled and attempt to sync bug information from specific related Bugzilla service.

This version works well, but not perfect. I'm being continue to add more features, such as the configuration ability.

### CODE REVIEW

File: GTG/backends/bugzilla/bugzilla.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        return [value]
+
+
+class BugInformationSyncTask(threading.Thread):
+
+    def __init__(self, task, backend, **kwargs):
+        ''' Initialize task data, where task is the GTG task object. '''
+        self.task = task
+        self.backend = backend
+        super(BugInformationSyncTask, self).__init__(**kwargs)
```
Comment (izidormatusov) [inline GTG/backends/bugzilla/bugzilla.py]: can you use python3 version of it?
```python
super().__init__(**kwargs)
```

File: GTG/backends/bugzilla/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+class BugzillaServiceNotExist(Exception):
+    pass
+
+
+class BugzillaServiceDisabled(Exception):
+    ''' Bugzilla service is disabled by user. '''
+
+    def __init__(self, domain, *args, **kwargs):
+        self.message = '%s is disabled.' % domain
+        super(BugzillaServiceDisabled, self).__init__(*args, **kwargs)
```
Comment (izidormatusov) [inline GTG/backends/bugzilla/exceptions.py]:
```python
super().__init__(*args, **kwargs)
```

File: GTG/gtk/backends_dialog/parameters_ui/ui_widget.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+__all__ = ('ParameterUIWidget',)
+
+
+class ParameterUIWidget(Gtk.Box):
+    '''Base class for definition of possible paramter UI'''
+
+    def __init__(self, req, backend, width,
+                 parameter_name, gtk_kwargs={}):
+        super(ParameterUIWidget, self).__init__(**gtk_kwargs)
```
Comment (izidormatusov) [inline GTG/gtk/backends_dialog/parameters_ui/ui_widget.py]:
```python
super().__init__(**gtk_kwargs)
```

**Comment by tkdchen** [comment]
@izidormatusov fixed according to your comments above.

**Comment by tkdchen** [comment]
strange. `make check` does not report any error locally in my machine. any idea about this result?

**Comment by izidormatusov** [comment]
Try to upgrade your pep8 command:

```
pip3 install -U pep8
```

File: GTG/backends/bugzilla/bug.py [hunk 4]
```diff
@@ -76,7 +64,17 @@ class RedHatBug(Bug):
 }
 
 
-class BugFactory(object):
-    @staticmethod
-    def create(serviceDomain, bug):
-        return bugs[serviceDomain](bug)
+def create_bug(service_domain, bug):
```
Comment (izidormatusov) [inline GTG/backends/bugzilla/bug.py]: What's the purpose of this whole module? To me it's enough to just check if  the service_domain is in the list of whitelisted domains, something like

```python

DOMAINS = set([
  'bugzilla.gnome.org',  'bugs.freedesktop.org', 'bugzilla.mozilla.org', 
  'bugzilla.samba.org', 'bugs.gentoo.org', 'bugzilla.redhat.com',
])
if service_domain not in DOMAINS:
  return None
```

File: GTG/backends/bugzilla/bugzilla.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            return
+
+        try:
+            bugzillaService = create_bugzilla_service(
+                task_info.scheme, task_info.net_location)
+        except (exceptions.BugzillaServiceNotExist,
+                exceptions.BugzillaServiceDisabled):
+            # Stop quietly when bugzilla cannot be found. Currently, I don't
+            # assume that user enters a wrong hostname or just an unkown
+            # bugzilla service.
```
Comment (izidormatusov) [inline GTG/backends/bugzilla/bugzilla.py]: Can you add at least `Log.warning()` to track when this happens?

File: GTG/backends/bugzilla/bugzilla.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+def parse_bug_url(url):
+    '''
+    Extract URL data
+
+    @param url: a string representing a URL
+    @return: a tuple containing scheme, hostname, and a list of key-value
+             queries
+    '''
+    r = urlparse(url)
+    queries = dict([item.split('=') for item in r.query.split('&')])
```
Comment (izidormatusov) [inline GTG/backends/bugzilla/bugzilla.py]: Do you mind to use https://docs.python.org/3.0/library/urllib.parse.html#urllib.parse.parse_qs instead?

File: GTG/backends/bugzilla/bugzilla.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        self.task = task
+        self.backend = backend
+        super().__init__(**kwargs)
+
+    def _collect_tags(self, bug):
+        tags = []
+        parameters = self.backend.get_parameters()
+
+        if parameters['bugzilla-tag-use-priority']:
+            tags += convert_to_list(bug.priority)
```
Comment (izidormatusov) [inline GTG/backends/bugzilla/bugzilla.py]: Why those are not just `[bug.priority]`?

File: GTG/backends/bugzilla/notification.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    Show notification in InfoBar
+
+    @param backend: the instance of Bugzilla backend
+    @param error_no: the error number defianed BackendSignals. Passing this
+                     follows the BackendSignals and CustomInfoBar protocol
+    @param error_message: the text message to be shown in InfoBar
+    '''
+    BackendSignals().backend_failed(backend.get_id(),
+                                    error_no,
+                                    {'message': error_message})
```
Comment (izidormatusov) [inline GTG/backends/bugzilla/notification.py]: I still think this should be moved to `BugInformationSyncTask`

File: GTG/tests/test_backend_bugzilla.py [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+
+    This method will be invoked in the separate process launched for running
+    the fake Bugzilla service.
+    '''
+
+    server = SimpleXMLRPCServer((XMLRPC_HOST, XMLRPC_PORT),
+                                logRequests=False,
+                                requestHandler=RequestHandler)
+    server.register_instance(FakeBugzillaService(), allow_dotted_names=True)
+    server.serve_forever()
```
Comment (izidormatusov) [inline GTG/tests/test_backend_bugzilla.py]: I don't like this unit test. Ideally you should split your code in that sense that it should be easy to mock the returned objects from the remote service like Bugzilla and test the rest of the object without spinning up a new XML RPC server. That would be slow & cumbersome and it wouldn't provide much benefit.

Have a look at [tests/core/test_config.py](https://github.com/getting-things-gnome/gtg/blob/master/tests/core/test_config.py) that does something similar: tests the loading up configuration from disk without actually touching the disk.

File: GTG/gtk/backends_dialog/parameters_ui/ui_widget.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+from gi.repository import Gtk
+
+__all__ = ('ParameterUIWidget',)
+
+
+class ParameterUIWidget(Gtk.Box):
+    '''Base class for definition of possible paramter UI'''
+
+    def __init__(self, req, backend, width,
+                 parameter_name, gtk_kwargs={}):
```
Comment (izidormatusov) [inline GTG/gtk/backends_dialog/parameters_ui/ui_widget.py]: make it `**kwargs`

File: GTG/gtk/backends_dialog/parameters_ui/importtagsui.py [hunk 11]
```diff
@@ -43,17 +46,14 @@ def __init__(self, req, backend, width, title, anybox_text, somebox_text,
...
-        self.backend = backend
-        self.req = req
         self.title = title
         self.anybox_text = anybox_text
         self.somebox_text = somebox_text
-        self.parameter_name = parameter_name
-        self._populate_gtk(width)
+        super(ImportTagsUI, self).__init__(
+            req, backend, width, parameter_name, gtk_kwargs={
+                'orientation': Gtk.Orientation.VERTICAL
```
Comment (izidormatusov) [inline GTG/gtk/backends_dialog/parameters_ui/importtagsui.py]: orientation=Gtk.Orientation.VERTICAL,

File: GTG/gtk/backends_dialog/parameters_ui/importtagsui.py [hunk 12]
```diff
@@ -43,17 +46,14 @@ def __init__(self, req, backend, width, title, anybox_text, somebox_text,
...
-        super(ImportTagsUI, self).__init__(
-            orientation=Gtk.Orientation.VERTICAL)
-        self.backend = backend
-        self.req = req
         self.title = title
         self.anybox_text = anybox_text
         self.somebox_text = somebox_text
-        self.parameter_name = parameter_name
-        self._populate_gtk(width)
+        super(ImportTagsUI, self).__init__(
```
Comment (izidormatusov) [inline GTG/gtk/backends_dialog/parameters_ui/importtagsui.py]: `super().__init__`

**Comment by izidormatusov** [comment]
@tkdchen Do you mind fixing those outlined comments?

**Comment by tkdchen** [comment]
I'll fix them all.

File: GTG/backends/bugzilla/bugzilla.py (hunk 7 above)
Comment (tkdchen) [inline GTG/backends/bugzilla/bugzilla.py]: Different Bugzilla service may return different attribution of a bug. For example, component of a bug returned from `bugzilla.redhat.com` is a list containing at least one component name, whereas it is a string value returned from `bugzilla.gnome.org`. Whatever a type of each attribution depends on the concrete Bugzilla service, and it is unpredictable for GTG or any other users.

`convert_to_list` is a defensive way to overcome this situation to the attributions used in Bugzilla synchronization service.

File: GTG/tests/test_backend_bugzilla.py (hunk 9 above)
Comment (tkdchen) [inline GTG/tests/test_backend_bugzilla.py]: This will be fixed in a separate commit.

**Comment by izidormatusov** [comment]
Hm... discussion on this PR is getting too old and too complicated. Do you mind to finish your work on this PR, merge master and fix tests?

We can fix the rest of the issues after this service is merged.

**Comment by tkdchen** [comment]
Dear @izidormatusov 

Sure. Agree with u. I'll merge master to this branch to solve all conflicts and push again.

**Comment by tkdchen** [comment]
@izidormatusov merged from master branch. please review.

**Comment by tkdchen** [comment]
What version of pyflake are you using to check code style locally? @izidormatusov

**Comment by izidormatusov** [comment]
Run

```
sudo make install
make check
```

**Comment by diegogangl** [comment]
@tkdchen are you still interested in working on this?

**Comment by tkdchen** [comment]
@diegogangl Currently, I'm not sure I can have enough time to maintain this and write more code for GTG. Feel free to take over it. Thanks.

**Comment by diegogangl** [comment]
We recently blacklisted the bugzilla plugin because of lack of manpower and none of us using it. I don't think we can also finish and maintain this. I'm closing for now, but we can leave a ticket open to finish this for 0.5 onwards.

@nekohayo What are your thoughts?
